plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -103,32 +103,52 @@ uriEncode(const String &in, bool isObjectName)
 }
 
 /**
- * @brief URI-decode a character string (AWS specific version, see spec)
+ * @brief checks if the string is URI-encoded (AWS specific encoding version, see spec)
  *
  * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
  *
- * @todo Consider reusing / converting to TSStringPercentDecode()
- *       Currently we don't build a library/archive so we could link with the unit-test binary. Also using
- *       different sets of encode/decode functions during runtime and unit-testing did not seem as a good idea.
- * @param in string to be URI decoded
- * @return encoded string.
+ * @note According to the following RFC if the string is encoded and contains '%' it should
+ *       be followed by 2 hexadecimal symbols otherwise '%' should be encoded with %25:
+ *          https://tools.ietf.org/html/rfc3986#section-2.1
+ *
+ * @param in string to be URI checked
+ * @param isObjectName if true encoding didn't encode '/', kept it as it is.
+ * @return true if encoded, false not encoded.
  */
-String
-uriDecode(const String &in)
+bool
+isUriEncoded(const String &in, bool isObjectName)
 {
-  std::string result;
-  result.reserve(in.length());
-  size_t i = 0;
-  while (i < in.length()) {
-    if (in[i] == '%') {
-      result += static_cast<char>(std::stoi(in.substr(i + 1, 2), nullptr, 16));
-      i += 3;
-    } else {
-      result += in[i];
-      i++;
+  for (size_t pos = 0; pos < in.length(); pos++) {
+    char c = in[pos];
+
+    if (isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
+      /* found a unreserved character which should not have been be encoded regardless
+       * 'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', and '~'.  */
+      continue;
+    }
+
+    if (' ' == c) {
+      /* space should have been encoded with %20 if the string was encoded */
+      return false;
+    }
+
+    if (isObjectName && '/' == c) {
+      /* if this is an object name '/' should have been encoded if the string was encoded */
+      return false;
+    }
+
+    if ('%' == c) {
+      if (pos + 2 <= in.length() && std::isxdigit(in[pos + 1]) && std::isxdigit(in[pos + 2])) {
```

Comment:
Shouldn't this be
```
if (pos + 2 <= in.length() && std::isxdigit(in[pos+1]) && std::isxdigit(in[pos+2] && (pos + 2 == in.length() || !std::isxdigit(in[pos+3])))
```

Reply:
If there is a third character that is a hex digit, we can interpret that as an un-encoded value. Having already seen we have 2 hex digits following a `%`, we know enough to return `true`, I think?

Reply:
@SolidWallOfCode and @d2r appreciate the rapid reviews! This check makes sure that there are at least 2 hexadecimal digits after the `%` before the end of the string, addressed and added extra relevant unit tests.

Reply:
No, because of your first statement. As you note, "%###" would not be un-encoded, therefore stopping short would be a false positive.

Reply:
> As you note, "%###" would not be encoded

I am not sure I follow to this statement. 

I would expect`%222` decodes to `"2` and therefore the `%222` string would be considered an encoded string: The first three characters, `%22` represent an encoded character `"`, and the final character `2` is un-encoded (i.e., it is literally just a `2` and not part of any encoding). 

If that's true so far, the result of evaluating this string should be `true` as it is a string with at least one encoded value, so that we can avoid double-encoding this string.

Reply:
@SolidWallOfCode, @d2r I wonder where the disconnect is, may be describing what I meant to do would help tracking it down (will use `XXX%123` as an example).
 
1.  A percent-encoded octet is encoded as a character triplet, consisting of the percent character `%` followed by the two hexadecimal digits representing that octet's numeric value.([RFC3986 section 2-1](https://tools.ietf.org/html/rfc3986#section-2.1))
1. In a properly precent-encoded string it would be impossible to encounter `%` followed by 0 or 1 hexadecimal digits since when encoding `%` would always end up encoded as `%25` ([RFC3986 section 2.4](https://tools.ietf.org/html/rfc3986#section-2.4))
1. We would need exactly `in[pos+0]='%'`, `in[pos+1]='1'` and `in[pos+2]='2'` to identify that the string was encoded. `in[pos+3]='3'` can be ignored.
1. To make sure we have at least 2 symbols to check after `%` and before the end of the string added a check for `pos + 2 < in.length()`. If we don't then the string is not (properly) encoded.

Reply:
@gtenev Agree, thanks.